src/ranged.cpp

```diff
@@ -376,6 +374,69 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
     return attack;
 }
 
+double player::gun_engagement_range( const item& gun, int aim, int penalty, unsigned chance, double accuracy ) const
+{
+    if( !gun.is_gun() || !gun.ammo_sufficient() ) {
+        return 0;
+    }
+
+    if( chance == 0 ) {
+        return gun.gun_range( this );
+    }
+
+    int driving = 0;
+    if( penalty < 0 ) {
+        // get current effective player recoil
+        penalty = std::max( recoil, 0 );
+        if( recoil < 0 ) {
+            debugmsg( "negative player recoil when calculating effective range" );
+        }
+
+        // get maximum penalty from driving
+        driving = std::max( driving_recoil, 0 );
+        if( driving_recoil < 0 ) {
+            debugmsg( "negative driving recoil when calculating effective range" );
+        }
+    }
+
+    // aim_per_time() returns improvement (in MOA) per 10 moves
+    for( int i = aim * 10; i != 0; --i ) {
+        int adj = aim_per_time( gun, penalty );
+        if( adj <= 0 ) {
+            break; // no further improvement is possible
+        }
+        penalty -= adj;
+    }
+
+    // calculate maximum potential dispersion
+    double dispersion = get_weapon_dispersion( &gun, false ) + penalty + driving;
+
+    // dispersion is uniformly distributed at random so scale linearly with chance
```

Comment:
Not sure if it is a good idea. It could work, but it would make OK shooters randomly miss completely and bad shooters much more likely to score perfect shots than now.